scripts/pylib/twister/twisterlib/runner.py

```diff
@@ -618,7 +619,11 @@ def process(self, pipeline, done, message, lock, results):
                     pipeline.put({"op": "report", "test": self.instance})
                 else:
                     logger.debug(f"Determine test cases for test instance: {self.instance.name}")
-                    self.determine_testcases(results)
+                    try:
+                        self.determine_testcases(results)
+                    except BuildError:
+                        self.instance.status = "error"
+                        self.instance.reason  = "Build Error"
                     pipeline.put({"op": "gather_metrics", "test": self.instance})
```

Comment:
I think this is not correct workflow. Note that `pipeline.put({"op": "gather_metrics", "test": self.instance})` is executed regardless if there was an error or not and the workflow continues with next steps. This can be seen if the issue dicussed on discord. if you have a test failing with an assert and get BuildError during a retry, with this implementation the test will proceed, call the execution and then replace the "error" status with the "failed" one.
I've seen you've accepted my PR, but since I had it commented, I am posting it for "educational" purpose.